reanahub/reana
# config: configurable job clean up on failure

* Closes #461.

The PR group to solve #461 is composed of:
- New Helm config to govern the feature: https://github.com/reanahub/reana/pull/484
- New central REANA-Commons config variable: https://github.com/reanahub/reana-commons/pull/249
- New central logic to determine whether to cleanup jobs: https://github.com/reanahub/reana-db/pull/121
- Usage of the new cleanup logic in RWC and RJC:
    - https://github.com/reanahub/reana-workflow-controller/pull/362
    - Second commit https://github.com/reanahub/reana-job-controller/pull/297
- Renaming of `succeeded` status inside RJC to `finished` following the [standard we set in REANA-DB](https://github.com/reanahub/reana-db/blob/a3adc6866ee3233e1c817825709979eb5b6165a2/reana_db/models.py#L291)
    - First commit in https://github.com/reanahub/reana-job-controller/pull/297
    - https://github.com/reanahub/reana-workflow-engine-yadage/pull/179
    - https://github.com/reanahub/reana-workflow-engine-serial/pull/133
    - https://github.com/reanahub/reana-workflow-engine-cwl/pull/169

Instructions to test:
- Deploy REANA with the new config variable, e.g.:
```diff
~/reana $ git diff                                                                                                                                                                                     [2:55:02]
diff --git a/helm/reana/values.yaml b/helm/reana/values.yaml
index 8ed0b7e..12a4ee7 100644
--- a/helm/reana/values.yaml
+++ b/helm/reana/values.yaml
@@ -69,6 +69,7 @@ components:
     image: reanahub/reana-workflow-controller:0.7.1
     environment:
       SHARED_VOLUME_PATH: /var/reana
+      REANA_KUBERNETES_JOBS_KEEP_ON_STATUS_LIST: failed,finished
   reana_workflow_engine_cwl:
     image: reanahub/reana-workflow-engine-cwl:0.7.2
   reana_workflow_engine_yadage:
```
- Run a workflow which will fail in some step, e.g.:
```diff
~/reana-demo-root6-roofit $ git diff
diff --git a/reana.yaml b/reana.yaml
index b1e8a2b..7fbdfac 100644
--- a/reana.yaml
+++ b/reana.yaml
@@ -18,7 +18,7 @@ workflow:
       - name: fitdata
         environment: 'reanahub/reana-env-root6:6.18.04'
         commands:
-        - root -b -q 'code/fitdata.C("${data}","${plot}")'
+        - unknown
 outputs:
   files:
     - results/plot.png
~/reana-demo-root6-roofit $ reana-client run
```
- Observe `kubectl get pods -w`
    - We should see that the workflow and its two jobs (one `failed`, one `finished`) are not cleaned up (or any other configured combination)
    - One can play changing `REANA_KUBERNETES_JOBS_KEEP_ON_STATUS_LIST` with other values for terminated workflows, for example with the current setup we should expect:
        - `REANA_KUBERNETES_JOBS_KEEP_ON_STATUS_LIST: failed`, just `fitdata` should remain
        - `REANA_KUBERNETES_JOBS_KEEP_ON_STATUS_LIST: finished`, just `gendata` should remain
       
       Note: one can identify if the job is gendata or fitdata from the logs: `kubectl logs reana-run-job-xxxx-yyy | grep "fitdata.C"` or `kubectl logs reana-run-job-xxxx-yyy | grep "gendata.C"`
